build: various cmake fixes

dundargoc opened the pull request:
- build: various cmake refactors and simplifications

Add STATUS keyword to message to ensure messages are shown in the
correct order.
Remove DEPS_CXX_COMPILER as we don't rely on C++ for any of our
dependencies.
Simplify how msgpack and luv configure options are constructed.
Rely on the default installation for luv instead of manually passing
configure, build and install commands.
Simplify return code conditional.

- build: remove CMAKE_OSX_ARCHITECTURES_ALT_SEP workaround

CMAKE_OSX_ARCHITECTURES_ALT_SEP was defined as a workaround to prevent
the shell from interpreting `;`, which CMake uses as a list separator.
However, the same thing can be achieved by instead passing
CMAKE_OSX_ARCHITECTURES as a cache variable instead, which is a more
idiomatic way of achieving the same thing.

- build: define CMAKE_BUILD_TYPE before adding it to BUILD_TYPE_STRING

The problem with the current setup is that CMAKE_BUILD_TYPE is defined
after BUILD_TYPE_STRING. BUILD_TYPE_STRING will then be empty on the
first run, meaning that dependencies are built without a build type.
However, since CMAKE_BUILD_TYPE is a cache variable its value will
persist in subsequent runs. On the second run BUILD_TYPE_STRING will
have the correct value, but it's a different value from the ones the
dependencies were built with. This will force some dependencies to be
built again.

Fixes https://github.com/neovim/neovim/issues/21672.

dundargoc committed:
build: various cmake refactors and simplifications

- Add STATUS keyword to message to ensure messages are shown in the
  correct order.
- Remove DEPS_CXX_COMPILER as we don't rely on C++ for any of our
  dependencies.
- Simplify how msgpack and luv configure options are constructed.
- Rely on the default installation for luv instead of manually passing
  configure, build and install commands.
- Simplify return code conditional.

dundargoc committed:
build: define CMAKE_BUILD_TYPE before adding it to BUILD_TYPE_STRING

The problem with the current setup is that CMAKE_BUILD_TYPE is defined
after BUILD_TYPE_STRING. BUILD_TYPE_STRING will then be empty on the
first run, meaning that dependencies are built without a build type.
However, since CMAKE_BUILD_TYPE is a cache variable its value will
persist in subsequent runs. On the second run BUILD_TYPE_STRING will
have the correct value, but it's a different value from the ones the
dependencies were built with. This will force some dependencies to be
built again.

Fixes https://github.com/neovim/neovim/issues/21672.

kylo252 commented on cmake.deps/cmake/BuildMsgpack.cmake:
```diff
@@ -1,23 +1,16 @@
+
+if(NOT MSVC)
+  list(APPEND MSGPACK_CMAKE_ARGS
+    -DCMAKE_OSX_ARCHITECTURES=${CMAKE_OSX_ARCHITECTURES_ALT_SEP}
```
```suggestion
    "-DCMAKE_OSX_ARCHITECTURES=${CMAKE_OSX_ARCHITECTURES_ALT_SEP}"
```

dundargoc replied:
?

kylo252 replied:
it's just a good habit when dealing with cmake to use quotes

dundargoc replied:
I'll take the risk if it means less cluttered cmake code ;)

kylo252 commented on cmake.deps/CMakeLists.txt:
```diff
@@ -12,13 +12,13 @@ list(APPEND CMAKE_MODULE_PATH "${PROJECT_SOURCE_DIR}/cmake" "${PROJECT_SOURCE_DI
+
 get_property(isMultiConfig GLOBAL PROPERTY GENERATOR_IS_MULTI_CONFIG)
 if(NOT isMultiConfig)
   set(BUILD_TYPE_STRING -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE})
```
this should be useful

```suggestion
  set(BUILD_TYPE_STRING -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE})
  message(STATUS "CMAKE_CONFIGURATION_TYPES=${CMAKE_CONFIGURATION_TYPES}")
```

dundargoc replied:
Wait, why?

kylo252 replied:
yeah, that's why I suggested to print it inside the condition when it's multi-config

dundargoc replied:
Sorry, I don't get it. The condition is `if(NOT isMultiConfig)`, so it won't be printed when it's multi-config?

clason commented on cmake.deps/CMakeLists.txt:
```diff
@@ -107,10 +107,6 @@ endif()
-if(CMAKE_OSX_SYSROOT)
-  set(DEPS_C_COMPILER "${DEPS_C_COMPILER} -isysroot${CMAKE_OSX_SYSROOT}")
-endif()
-
```
Nope. That breaks dependency builds.

dundargoc replied:
Sure, just wanna check the errors :)

dundargoc committed:
build: remove CMAKE_OSX_ARCHITECTURES_ALT_SEP workaround

CMAKE_OSX_ARCHITECTURES_ALT_SEP was defined as a workaround to prevent
the shell from interpreting `;`, which CMake uses as a list separator.
However, the same thing can be achieved by instead passing
CMAKE_OSX_ARCHITECTURES as a cache variable instead, which is a more
idiomatic way of achieving the same thing.

dundargoc commented:
I was wondering if anyone with a mac, preferably an M1, can try out this PR and see if it works. I'm pretty confident that the workaround I removed shouldn't be needed, but it can't hurt to be safe.

clason commented:
@carlocab (who I believe contributed the `CMAKE_OSX_ARCHITECTURES_ALT_SEP` thing to make our universal release workflow work)

dundargoc commented:
Tested release job: https://github.com/dundargoc/neovim/actions/runs/3968011322/jobs/6800640437

lewis6991 commented:
Works for me.
